File: week01/example/ltc20_valid_parentheses.cpp
//
// Created by Opt_c on 2022/4/18.
//
#include <bits/stdc++.h>
#include <cassert>

using namespace std;


class Solution {
public:
    bool isValid(string s) {
        for (char ch: s) {
            if (ch == '(' || ch == '[' || ch == '{') a.push(ch);
            else {
                if (a.empty()) return false;
                if (ch == ')' && '(' != a.top()) return false;
                if (ch == ']' && '[' != a.top()) return false;
                if (ch == '}' && '{' != a.top()) return false;
                a.pop();
            }
        }
        return a.empty();
    }

private:
    stack<char> a;
};


int main() {
    assert(Solution().isValid("()") == true);
    assert(Solution().isValid("{[]}") == true);
    assert(Solution().isValid("([)]") == false);
    return 0;
}

File: week01/example/ltc20_valid_parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = list()
        for i in s:
            if i == '(' or i == '[' or i == '{':
                stack.append(i)
            else:
                if len(stack) == 0:
                    return False
                if (i == ')' and stack[-1] != '(') \
                        or (i == ']' and stack[-1] != '[') \
                        or (i == '}' and stack[-1] != '{'):
                    return False
                stack.pop()
        return len(stack) == 0


if __name__ == '__main__':
    str_list = ["()", "()[]{}", "(]", "([)]", "{[]}"]
    expect_res = [True, True, False, False, True]
    for i in range(len(str_list)):
        assert Solution().isValid(str_list[i]) == expect_res[i]
